crates/engine/tree/src/tree/root.rs

```diff
@@ -464,31 +464,50 @@ where
     }
 }
 
+/// Returns accounts only with those storages that were not already fetched, and
+/// if there are no such storages and the account itself was already fetched, the
+/// account shouldn't be included.
 fn get_proof_targets(
     state_update: &HashedPostState,
     fetched_proof_targets: &HashMap<B256, HashSet<B256>>,
 ) -> HashMap<B256, HashSet<B256>> {
-    state_update
-        .accounts
-        .keys()
-        .filter(|hashed_address| !fetched_proof_targets.contains_key(*hashed_address))
-        .map(|hashed_address| (*hashed_address, HashSet::default()))
-        .chain(state_update.storages.iter().map(|(hashed_address, storage)| {
-            let fetched_storage_proof_targets = fetched_proof_targets.get(hashed_address);
-            (
-                *hashed_address,
-                storage
-                    .storage
-                    .keys()
-                    .filter(|slot| {
-                        !fetched_storage_proof_targets
-                            .is_some_and(|targets| targets.contains(*slot))
-                    })
-                    .copied()
-                    .collect(),
-            )
-        }))
-        .collect()
+    let mut targets = HashMap::default();
+
+    // first collect all new accounts (not previously fetched)
+    for &hashed_address in state_update.accounts.keys() {
+        if !fetched_proof_targets.contains_key(&hashed_address) {
+            targets.insert(hashed_address, HashSet::default());
+        }
+    }
+
+    // then process storage slots for all accounts in the state update
+    for (hashed_address, storage) in &state_update.storages {
+        // only process storage if we either:
+        // 1. haven't fetched this account at all (it's in our targets), or
+        // 2. have fetched the account but need new storage slots
+        if targets.contains_key(hashed_address) ||
+            fetched_proof_targets.contains_key(hashed_address)
+        {
+            let target_slots = targets.entry(*hashed_address).or_default();
+            let fetched_storage_slots = fetched_proof_targets.get(hashed_address);
+
+            // add only storage slots that haven't been fetched yet
+            for slot in storage.storage.keys() {
+                if !fetched_storage_slots.is_some_and(|fetched_slots| fetched_slots.contains(slot))
+                {
+                    target_slots.insert(*slot);
+                }
+            }
```

Comment:
can collapse it into a one `if let Some(fetched_storage_slots) = fetched_proof_targets.get(hashed_address) { ... } else { ... }` to do less branching on `fetched_storage_slots.is_some_and`

Reply:
good point done